### DISCUSSION TITLE
Does a control or its children still render when isVisibile is False?

### DISCUSSION THREAD

**Discussion by Tamnac** [AvaloniaUI/Avalonia#12517] [Q&A] [answered]
What I'm trying to ask is, if I have a big old control in the xaml, but I set `isVisible` to `False`, does it:
 1. Still do the work of layout and rendering, but just not display in the final result
 2. Not do any work at all

(same goes for children)

I'm asking because if I have some heavy control with a lot of items, but I remove it using `isVisible` would there still be a performance hit?

Thanks

**Comment by timunie** [comment]
The workload should at least be less as the control should not render at all. But as you may have some custom logic I don't know, I'd suggest to create a minimal sample and try to see what happens if you toggle visibility of your complex control. You should see some performance improvements if it's hidden.

**Reply by Tamnac** [reply]
I don't really have a working example. It was moreso a conceptual question at this point, as I'm a beginner and don't have much knowledge of avalonia or gui programming in general. Thanks for your help though

**Comment by robloo** [comment, marked as answer]
A little history here. WPF had the Visibility property with 3 possible values: Visible, Hidden and Collapsed. Hidden will still measure and layout the control -- it just won't render it. Hidden controls still take up space in the UI and affect the position of controls around them. Collapsed controls, however, are not used in layout or shown in the render pass, it's as if they just completely disappeared.

Now in hindsight even Microsoft's original engineers admit having an enum with 3 values was a mistake. In fact in UWP/WinUI they went back to 2 values: Visible and Collapsed (for XAML compatibility). This is because Hidden is equivalent to just setting opacity to zero (Transparent controls still are used in layout but are not visible after rendering).

In Avalonia IsVisible maps to the original WPF terminology as below:

 IsVisible = True : Visible
 IsVisible = False : Collapsed

This means setting IsVisible to false will NOT use the control for layout and NOT use the control for rendering. It will just disappear. So the answer is number 2 in your list.

Now I'm not 100% sure Avalonia behaves this way, but I would expect the measure and arrange methods to never get called on a custom control that has IsVisible set to false. [Edit: As clarified below the methods do get called but return zero size] It definitely won't be used in the renderer though. You might want to test that out and there could be something there that needs to be looked at in more detail (I'm not sure there is a unit test for this or if anyone has checked).

Finally, usually controls should be aware of either: Loaded/Unloaded state or Attached/DetachedFromVisualTree. They use these methods or events to disable themselves and cleanup when needed.

**Reply by Tamnac** [reply]
Thank you! That's very helpful

**Reply by binarycow** [reply]
> Now I'm not 100% sure Avalonia behaves this way, but I would expect the measure and arrange methods to never get called on a custom control that has IsVisible set to false.

The default implementation of `Layoutable.MeasureCore` is to return `new Size()` if the control's visibility is `false`

[Source](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Layout/Layoutable.cs#L524)
